### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by Maxtaager**
generic search please review @KristianOellegaard

### CODE REVIEW

File: smart_lists/helpers.py [hunk 1]
```diff
@@ -226,6 +226,8 @@ def __init__(self, object_list, list_settings):
         self.query_params = list_settings.get('query_params', {})
         self.list_display = list_settings.get('list_display')
         self.list_filter = list_settings.get('list_filter')
+        self.search_fields = list_settings.get('search_fields')
+        self.search_term = list_settings.get('search_term')
```
Comment (KristianOellegaard) [inline smart_lists/helpers.py]: This should be consistently named with ```self.ordering_query_param``` - I suggest ```self.search_query_param```

File: smart_lists/mixins.py [hunk 2]
```diff
@@ -24,8 +27,40 @@ def get_queryset(self):
         if filters:
             for fltr in filters:
                 qs = qs.filter(**fltr)
+        qs = self.get_search_results(qs)
```
Comment (KristianOellegaard) [inline smart_lists/mixins.py]: Use similar code style as filters and ordering

```
search = self.get_search()
if search:
    qs = qs.filter(**search)
```

File: smart_lists/mixins.py [hunk 3]
```diff
@@ -24,8 +27,40 @@ def get_queryset(self):
...
+        borrowed from django-admin
+        @param queryset:
+        @return: filtered queryset
+        """
+        search_term = self.request.GET.get(self.search_query_parameter_name, None)
+
+        if search_term is None:
+            return queryset
+        # Apply keyword searches.
+        def construct_search(field_name):
```
Comment (KristianOellegaard) [inline smart_lists/mixins.py]: Don't define functions inside class methods

File: smart_lists/mixins.py [hunk 4]
```diff
@@ -24,8 +27,40 @@ def get_queryset(self):
...
+            if field_name.startswith('^'):
```
Comment (KristianOellegaard) [inline smart_lists/mixins.py]: Re-implement search lookup names similar to django admin, not use strange looking characters that might have weird URL encodings.

Query should look like this:
```?q=asd```

Variables should look like this:
```search_fields = ('company_name', 'code')```

**Review by KristianOellegaard — CHANGES_REQUESTED** [review]
Hey Mikkel,

Please take a look at these changes and then I think we can merge it in.

Thank you very much for the work so far on this!

File: smart_lists/mixins.py (hunk 4 above)
Comment (Maxtaager) [reply smart_lists/mixins.py]: i did take it from django-admin django/contrib/admin/options.py class ModelAdmin method get_search_result.
A crazy idea would be to call this method directly... (but that might break across django versions).
def get_search_results(self, request, queryset, search_term):

or i could rewrite the code using a more simple lookup with only contains, that would make the code much cleaner.

Comment (Maxtaager) [reply smart_lists/mixins.py]: renamed search parameter to q=? instead of search=?

Comment (KristianOellegaard) [reply smart_lists/mixins.py]: LGTM

**Pull request closed without merging**